src/bridge/cockpitrouter.c

```diff
@@ -323,6 +364,146 @@ process_open_peer (CockpitRouter *self,
   return cockpit_peer_handle (peer, channel, options, data);
 }
 
+static GBytes *
+substitute_json_string (const gchar *variable,
+                        gpointer user_data)
+{
+  const gchar *value;
+  JsonObject *options = user_data;
+  if (options && cockpit_json_get_string (options, variable, "", &value))
+    return g_bytes_new (value, strlen (value));
+  else if (options)
+    g_message ("Couldn't get argument for bridge: got invalid value for '%s'", variable);
+
+  return g_bytes_new ("", 0);
+}
+
+static JsonArray *
+ptr_array_to_json_array (GPtrArray *arr,
+                         gint start,
+                         gint end)
+{
+  JsonArray *new;
+  gint i;
+
+  g_return_val_if_fail (arr != NULL, NULL);
+
+  new = json_array_new ();
+  if (end > arr->len)
+    end = arr->len;
+
+  for (i = start; i < end; i++)
+    {
+      const gchar *d = g_ptr_array_index (arr, i);
+      g_assert (d != NULL);
+      json_array_add_string_element (new, d);
+    }
+
+  return new;
+}
+
+static void
+add_dynamic_args_to_array (GPtrArray *arr,
+                           gchar **config_args,
+                           JsonObject *options)
+{
+  gint length;
+  gint i;
+
+  g_return_if_fail (config_args != NULL);
+  g_return_if_fail (arr != NULL);
+
+  length = g_strv_length (config_args);
+  for (i = 0; i < length; i++)
+    {
+      GString *s = g_string_new ("");
+      GBytes *input = g_bytes_new_static (config_args[i], strlen(config_args[i]) + 1);
+      GList *output = cockpit_template_expand (input, substitute_json_string,
+                                               "${", "}", options);
+      GList *l;
+      for (l = output; l != NULL; l = g_list_next (l))
+        {
+          gsize size;
+          gconstpointer data = g_bytes_get_data (l->data, &size);
+          g_string_append_len (s, data, size);
+        }
+
+      g_ptr_array_add (arr, g_string_free (s, FALSE));
+      g_bytes_unref (input);
+      g_list_free_full (output, (GDestroyNotify) g_bytes_unref);
+    }
+}
+
+static gboolean
+process_open_dynamic_peer (CockpitRouter *self,
+                           const gchar *channel,
+                           JsonObject *options,
+                           GBytes *data,
+                           gpointer user_data)
+{
+  CockpitPeer *peer = NULL;
+  DynamicPeer *dp = user_data;
+  JsonObject *config = NULL;
+  GPtrArray *arr = NULL;
+  gchar *key = NULL;
+  gint env_spot = 0;
+
+  arr = g_ptr_array_new_with_free_func (g_free);
+  if (dp->spawn)
+    add_dynamic_args_to_array (arr, dp->spawn, options);
```

Comment:
I don't see how the split in key between the argv and env happens. Couldn't a trailing value in ```spawn``` be conflated with a leading value in ```environ```?